towhee/hub/download_tools.py

```diff
@@ -0,0 +1,251 @@
+    """
+
+    url = f'https://hub.towhee.io/api/v1/repos/{author}/{repo}/commits?limit=1&page=1&sha={tag}'
+    try:
+        r = requests.get(url, allow_redirects=True)
+        r.raise_for_status()
+    except HTTPError as e:
+        raise e
+
+    res = r.json()
+
+    return res[0]['sha'][:10]
+
+
+def get_file_list(author: str, repo: str, commit: str) -> List[str]:
+    """
+    Get all the files in the current repo at the given commit.
+
+    This is done through forming a git tree recursively and filtering out all the files.
+
+    Args:
+        author (`str`):
+            The account name.
+        repo (`str`):
+            The repo name.
+        commit (`str`):
+            The commit to base current existing files.
+
+    Returns:
+        (`List[str]`)
+            The file paths for the repo
+
+    Raises:
+        (`HTTPError`)
+            Raise error in request.
+    """
+
+    url = f'https://hub.towhee.io/api/v1/repos/{author}/{repo}/git/trees/{commit}?recursive=1'
+    file_list = []
+    try:
+        r = requests.get(url)
+        r.raise_for_status()
+    except HTTPError as e:
+        raise e
+
+    res = r.json()
+    # Check each object in the tree
+    for file in res['tree']:
+        # Ignore directories (they have the type 'tree')
+        if file['type'] != 'tree':
+            file_list.append(file['path'])
+
+    return file_list
+
+
+def download_files(author: str, repo: str, tag: str, file_list: List[str], lfs_files: List[str], local_dir: str, install_reqs: bool) -> None:
+    """
+    Download the files from hub. One url is used for git-lfs files and another for the other files.
+
+    Args:
+        author (`str`):
+            The account name.
+        repo (`str`):
+            The repo name.
+        tag (`str`):
+            The tag name.
+        file_list (`List[str]`):
+            The hub file paths.
+        lfs_files (`List[str]`):
+            The file extensions being tracked by git-lfs.
+        local_dir (`str`):
+            The local directory to download to.
+        install_reqs (`bool`):
+            Whether to install packages from requirements.txt
+
+    Raises:
+        (`HTTPError`)
+            Rasie error in request.
+        (`OSError`)
+            Raise error in writing file.
+    """
+    threads = []
+
+    # If the trailing forward slash is missing, add it on.
+    if local_dir[-1] != '/':
+        local_dir += '/'
+
+    # endswith() can check multiple suffixes if they are a tuple.
+    lfs_files = tuple(lfs_files)
+
+    for file_name in file_list:
+        # Files dealt with lfs have a different url.
+        if file_name.endswith(lfs_files):
+            url = f'https://hub.towhee.io/{author}/{repo}/media/branch/{tag}/{file_name}'
+        else:
+            url = f'https://hub.towhee.io/api/v1/repos/{author}/{repo}/raw/{file_name}?ref={tag}'
+
+        threads.append(Worker(url, local_dir, file_name))
+        threads[-1].start()
+
+    for thread in threads:
```

Comment:
If a thread has errors, will it stop? Because the exception is raised in another thread.